nvflare/private/fed/client/client_executor.py

```diff
@@ -317,24 +317,38 @@ def send_aux_command(self, shareable: Shareable, run_number):
 
     def abort_train(self, client, run_number):
         with self.lock:
-            process_status = self.run_processes.get(run_number, {}).get(RunProcessKey.STATUS, ClientStatus.NOT_STARTED)
-            if process_status == ClientStatus.STARTED:
-                try:
-                    child_process = self.run_processes[run_number][RunProcessKey.CHILD_PROCESS]
-                    conn_client = self.get_conn_client(run_number)
-                    if conn_client:
-                        data = {"command": AdminCommandNames.ABORT, "data": {}}
-                        conn_client.send(data)
-                        self.logger.debug("abort sent")
-
-                    threading.Thread(target=self._terminate_process, args=[child_process, run_number]).start()
-                except Exception as e:
-                    self.logger.error(f"abort_train execution exception: {e}.", exc_info=True)
-                finally:
-                    if conn_client:
-                        conn_client.close()
-                    self.run_processes.pop(run_number)
-                    self.cleanup()
+            # When the HeartBeat cleanup process try to abort the train, the job maybe already terminated,
+            # Use retry to avoid print out the error stack trace.
+            retry = 1
+            while retry >= 0:
+                process_status = self.run_processes.get(run_number, {}).get(
+                    RunProcessKey.STATUS, ClientStatus.NOT_STARTED
+                )
+                if process_status == ClientStatus.STARTED:
+                    try:
+                        child_process = self.run_processes[run_number][RunProcessKey.CHILD_PROCESS]
+                        conn_client = self.get_conn_client(run_number)
+                        if conn_client:
+                            data = {"command": AdminCommandNames.ABORT, "data": {}}
+                            conn_client.send(data)
+                            self.logger.debug("abort sent")
+
+                        threading.Thread(target=self._terminate_process, args=[child_process, run_number]).start()
+                        break
+                    except Exception as e:
+                        if retry == 0:
+                            self.logger.error(
+                                f"abort_train execution exception: {e} for run: {run_number}.", exc_info=True
+                            )
+                        retry -= 1
+                        time.sleep(5.0)
+                    finally:
+                        if conn_client:
+                            conn_client.close()
+                        self.run_processes.pop(run_number)
+                        self.cleanup()
```

Comment:
This is intended, which means the run already terminated. In this case, no need to run the ABORT again.

Reply:
If the process status is not `ClientStatus.STARTED`
This code will never end since `retry` is not being updated.

Reply:
If at first the process status is `ClientStatus.STARTED`

some exception happens, retry will be 0, then the process will be pop.
then this while loop never ends.

Reply:
Addressed this issue. Thanks for catching it!

Reply:
No, in normal case, when we retry the next time, the process_status will be NOT_STARTED, because the process already terminated. But the error logging is to handle the exception case if it's still STARTED, and we got the exception.

Reply:
Changed self.run_processes.pop(run_number) in the right place.